November 5 2024 - Adding spotlight for Jesse Glick (#214)
jenkins-infra/contributor-spotlight: src/pages/contributors/jesse-glick.adoc

= Jesse Glick
:page-name: Jesse Glick
:page-linkedin:
:page-twitter: 
:page-github: jglick
:page-email:
:page-image: avatar/jesse-glick.jpeg
:page-pronouns:
:page-location: Raleigh, North Carolina, USA
:page-firstcommit: 2009
:page-datepublished: 2024-11-05
:page-featured: true

> :page-intro: Jesse Glick is a software engineer, currently residing in Raleigh, North Carolina and has been involved with the Jenkins project for over a decade. Prior to contributing to Jenkins, Jesse worked as an engineer at Sun Microsystems. After incorporating Jenkins to their work, albeit in a small way, there wasn't much more to do with the project. However, that all changed when Kohsuke reached out to Jesse to find out if he would be interested in working on Jenkins full time. Since joining CloudBees in 2012, Jesse has been an integral part of Pipeline development and incorporating many functions that are now fundamental to current Pipeline operation. Jesse is constantly working on improving the Pipeline experience and shares his insights on how to improve not only the Jenkins project, but the open-source workflow as a whole.

Reviewer: ```suggestion
:page-intro: Jesse Glick is a software engineer, currently residing in Raleigh, North Carolina and has been involved with the Jenkins project for over a decade. Prior to contributing to Jenkins, Jesse worked as an engineer at Sun Microsystems. After incorporating Jenkins to his work at Sun, albeit in a small way, there wasn't much more to do with the project. However, that all changed when Kohsuke reached out to Jesse to find out if he would be interested in working on Jenkins full time. Since joining CloudBees in 2012, Jesse has been an integral part of Pipeline development, incorporating many functions that are now fundamental to current Pipeline operation. Jesse works constantly on improving the Pipeline experience, and shares his insights on how to improve not only the Jenkins project, but the open-source workflow as a whole.
```

> :page-intro: Jesse Glick is a software engineer, currently residing in Raleigh, North Carolina and has been involved with the Jenkins project for over a decade. Prior to contributing to Jenkins, Jesse worked as an engineer at Sun Microsystems. After incorporating Jenkins to his work at Sun, albeit in a small way, there wasn't much more to do with the project. However, that all changed when Kohsuke reached out to Jesse to find out if he would be interested in working on Jenkins full time. Since joining CloudBees in 2012, Jesse has been an integral part of Pipeline development, and incorporating many functions that are now fundamental to current Pipeline operation. Jesse works constantly on improving the Pipeline experience, and shares his insights on how to improve not only the Jenkins project, but the open-source workflow as a whole.

Reviewer 2: ```suggestion
:page-intro: Jesse Glick is a software engineer, currently residing in Raleigh, North Carolina, and has been involved with the Jenkins project for over a decade. Prior to contributing to Jenkins, Jesse worked as an engineer at Sun Microsystems. After incorporating Jenkins into his work at Sun, albeit in a small way, there wasn't much more to do with the project. However, that all changed when Kohsuke reached out to Jesse to find out if he would be interested in working on Jenkins full-time. Since joining CloudBees in 2012, Jesse has been an integral part of Pipeline development, and incorporating many functions that are now fundamental to current Pipeline operation. Jesse works constantly on improving the Pipeline experience and shares his insights on how to improve not only the Jenkins project but the open-source workflow as a whole.
```

> :page-intro: Jesse Glick is a software engineer, currently residing in Raleigh, North Carolina, and has been involved with the Jenkins project for over a decade. Prior to contributing to Jenkins, Jesse worked as an engineer at Sun Microsystems. After incorporating Jenkins into his work at Sun, albeit in a small way, there wasn't much more to do with the project. However, that all changed when Kohsuke reached out to Jesse to find out if he would be interested in working on Jenkins full-time. Since joining CloudBees in 2012, Jesse has been an integral part of Pipeline development, and incorporating many functions that are now fundamental to current Pipeline operation. Jesse works constantly on improving the Pipeline experience, and shares his insights on how to improve not only the Jenkins project but the open-source workflow as a whole.

Reviewer 3: ```suggestion
:page-intro: Jesse Glick is a software engineer, currently residing near Raleigh, North Carolina, and has been involved with the Jenkins project for over a decade. Prior to contributing to Jenkins, Jesse worked as an engineer at Sun Microsystems. After incorporating Jenkins into his work at Sun, albeit in a small way, there wasn't much more to do with the project. However, that all changed when Kohsuke reached out to Jesse to find out if he would be interested in working on Jenkins full-time. Since joining CloudBees in 2012, Jesse has been an integral part of Pipeline development, and incorporating many functions that are now fundamental to current Pipeline operation. Jesse works constantly on improving the Pipeline experience, and shares his insights on how to improve not only the Jenkins project but the open-source workflow as a whole.
```

BTW https://asciidoctor.org/docs/asciidoc-recommended-practices/#one-sentence-per-line makes reviews in GH a lot more pleasant.

Author: @jglick I definitely agree with the one sentence per line, and that's typically what we use. The issue with doing that for the page intro is that it breaks it up completely and doesn't render as desired on the overview page. The goal for the intro is to force an `...` within the feature box to indicate there is more to read. Doing one sentence per line would make it so only the first line included in the tag is displayed and would look inadequate.
<img width="577" alt="Screenshot 2024-10-30 at 12 54 46 PM" src="https://github.com/user-attachments/assets/1f94fb74-5ab1-4a76-9eab-616570d45194">

> :page-intro: Jesse Glick is a software engineer, currently residing in Raleigh, North Carolina and has been involved with the Jenkins project for over a decade. Prior to contributing to Jenkins, Jesse worked as an engineer at Sun Microsystems. After incorporating Jenkins to his work at Sun, albeit in a small way, there wasn't much more to do with the project. However, that all changed when Kohsuke reached out to Jesse to find out if he would be interested in working on Jenkins full time. Since joining CloudBees in 2012, Jesse has been an integral part of Pipeline development, and incorporating many functions that are now fundamental to current Pipeline operation. Jesse works constantly working on improving the Pipeline experience, and shares his insights on how to improve not only the Jenkins project, but the open-source workflow as a whole.

Reviewer 4: ```suggestion
:page-intro: Jesse Glick is a software engineer, currently residing in Raleigh, North Carolina and has been involved with the Jenkins project for over a decade. Prior to contributing to Jenkins, Jesse worked as an engineer at Sun Microsystems. After incorporating Jenkins to his work at Sun, albeit in a small way, there wasn't much more to do with the project. However, that all changed when Kohsuke reached out to Jesse to find out if he would be interested in working on Jenkins full time. Since joining CloudBees in 2012, Jesse has been an integral part of Pipeline development, and incorporating many functions that are now fundamental to current Pipeline operation. Jesse works constantly on improving the Pipeline experience, and shares his insights on how to improve not only the Jenkins project, but the open-source workflow as a whole.
```

== What is your background prior to contributing to Jenkins?

> I worked at Sun Microsystems for quite a long time, working on multiple projects including NetBeans ID.

Reviewer 3: ```suggestion
I worked at Sun Microsystems for quite a long time, working on multiple projects including NetBeans IDE.
```

We had some CI processes, such as Cruise Control, as part of that.
While there, I heard about Jenkins (then Hudson) in its infancy.
I knew of Kohsuke and his work on Hudson and thought it might be something we could start using.
At the time, we had been using CVS for version control, which Hudson had incorporated to some degree.

> We started using Hudson enough that we transitioned from CVS to Mercurial SCM, which I also wound up working on in Hudson.

Reviewer 3: Non sequitur.

```suggestion
We migrated from CVS to Mercurial SCM, which I also wound up working on in Hudson.
```

== Why choose Jenkins over other projects?

There weren't a lot of options at the time, with Cruise Control being the only other alternative.
It worked and was very extensible.
Jenkins either did what we needed it to or we could add plugins that would perform the necessary functions.
After becoming familiar with the system, extending Jenkins further was fairly straightforward.
For instance, we wanted to generate a historical diff for quick reference that anyone could use.
This wasn't possible inherently with Jenkins, but we were able to create a plugin that completed this particular task.
Jenkins' versatility has only expanded since then, proving to be one of its greatest strengths.

== What problems have you solved in Jenkins?

I'm always looking for opportunities to simplify some parts of the architecture where possible.
One project I worked on was the re-design of the Jenkins CLI.
At the time, it was very troublesome from a security perspective and we would have to issue emergency advisories for something like the controller sending back something malicious.

> The project goal was to rip out all of the problematic functionality and turn it into a dump client.

Reviewer 3: ```suggestion
The project goal was to rip out all of the problematic functionality and turn it into a dumb client.
```

It could send a command and get back a stream of text, and the protocol was simplified so that it couldn't do anything more than these basic functions.
However, this also meant tracking down every plugin that might interact with the CLI in this way and either updating code as needed or deprecating the function so that this would not persist.

> I also did some work on plugable storage changes, mainly the link:https://plugins.jenkins.io/artifact-manager-s3/[Artifact Manager on S3] plugin, to help externalize the storage of artifacts and statuses.

Reviewer 3: ```suggestion
I also did some work on pluggable storage changes, mainly the link:https://plugins.jenkins.io/artifact-manager-s3/[Artifact Manager on S3] plugin, to help externalize the storage of artifacts and stashes.
```

> I also did some work on plugable storage changes, mainly the Artifact Manager on S3 plugin, to help externalize the storage of artifacts and statuses.
> The plugin ecosystem also forces you to work in particular ways, such as keeping track of what plugins might be affected by a change, or determining how a plugin might interact with various functions or other plugins. 
> It also pushes us to track down plugin maintainers for assistance with changes or updates, to ensure things are not broken by Jenkins core updates.

Reviewer 2: ```suggestion
I also did some work on plugable storage changes, mainly the Artifact Manager on the S3 plugin, to help externalize the storage of artifacts and statuses.
The plugin ecosystem also forces you to work in particular ways, such as keeping track of what plugins might be affected by a change, or determining how a plugin might interact with various functions or other plugins. 
It also pushes us to track down plugin maintainers for assistance with changes or updates, to ensure things are not broken by Jenkins core updates.
```

Author: The plugin name is Artifact Manager on S3, https://plugins.jenkins.io/artifact-manager-s3/. I'm happy to make this change for the grammar aspect, but it isn't correct according to the plugin page.

Reviewer 2: My bad, thanks for your insight, Kevin. 👍

Author: No worries! I believe it came up at another point and I didn't clarify, so I figured I would do so and add a link to the plugin itself just to cover all bases.

== Is there an aspect of Jenkins that you're particularly interested in?

A lot of my work has been done on Pipeline so I am partial to that.
There have been several requests or bug reports from users that have also driven the work we do within Pipeline and Jenkins as a whole.
These requests also help drive architectural changes, despite the challenges they can present.
Such requests also help find and deal with technical debt.

== What contributions have felt the most successful or impactful?

> The changes and advancements in Pipeline, including Multibranch Pipeline and Scripted Pipeline.
> 
> // Left mostly blank for Jesse to provide additional information/response

Reviewer 3: ```suggestion
I was eager to file one of the first substantive link:https://github.com/jenkinsci/jep/tree/master/jep[JEPs]:
JEP-200, greatly reducing the surface area of Jenkins threatened by deserialization “gadgets”
(in the same spirit as the Jenkins CLI simplification, which predated the JEP system).
Another notable feature was JEP-222, allowing WebSocket to be used to connect agents to the controller.
This was inspired by the trickiness of setting up “level 4” network proxies,
a common hurdle for CloudBees CI customers especially in Kubernetes.

JEP-227 was a very tedious, yet long overdue, project to update to the (then-) current Spring Security release across several major versions.
Preventing almost all user-visible regressions required a lot of refactoring in Jenkins core and a lot of plugin updates.
It is great to see that the Jenkins project has gotten much more proactive in updating dependencies since those days.

At the infrastructure level, JEP-305 made it possible to prototype changes spanning multiple plugins without sacrificing CI coverage,
despite the lack of a “monorepo” for Jenkins development.
More recently, JEP-229 built on that system to make it possible to publish plugin releases automatically
without requiring maintainers to keep Artifactory credentials on their laptop,
a longstanding source of frustrating mistakes (and a security risk).
Both changes were inspired especially by work on Pipeline-related plugins,
which form a loose coalition without a fixed boundary:
I wanted it to be as simple as possible to develop, test, and publish changes rapidly.
```

= Jesse Glick
:page-name: Jesse Glick
:page-linkedin:
:page-twitter: 
:page-github: jglick
:page-email:
:page-image: avatar/jesse-glick.jpeg
:page-pronouns:
:page-location: Raleigh, North Carolina, USA
:page-firstcommit: 2009
:page-datepublished: 2024-11-05
:page-featured: true
:page-intro: Jesse Glick is a software engineer, currently residing near Raleigh, North Carolina, and has been involved with the Jenkins project for over a decade. Prior to contributing to Jenkins, Jesse worked as an engineer at Sun Microsystems. After incorporating Jenkins into his work at Sun, albeit in a small way, there wasn't much more to do with the project. However, that all changed when Kohsuke reached out to Jesse to find out if he would be interested in working on Jenkins full-time. Since joining CloudBees in 2012, Jesse has been an integral part of Pipeline development, and incorporating many functions that are now fundamental to current Pipeline operation. Jesse works constantly on improving the Pipeline experience, and shares his insights on how to improve not only the Jenkins project but the open-source workflow as a whole.

== What is your background prior to contributing to Jenkins?

> I worked at Sun Microsystems for quite a long time, working on the NetBeans IDE.
> We had some CI processes, such as Cruise Control, as part of that.
> While there, I heard about Jenkins (then Hudson) in its infancy.
> I knew of Kohsuke and his work on Hudson and thought it might be something we could start using.
> At the time, we had been using CVS for version control, which Hudson had incorporated to some degree.
> 
> We migrated from CVS to Mercurial SCM, which I also wound up working on in Hudson.
> It was just enough to get the features we needed for our usage patterns and we didn't pay more attention to it beyond that.
> I had also done some work on the tool system where you can have tool installations and automatic downloads.
> That was mostly driven by Andrew Bayer, another long-time Jenkins contributor.
> After the Jenkins fork was created, Kohsuke contacted me and asked if I wanted to come join him at the company where he was working on Jenkins.
> This was in 2012 when I joined CloudBees, and I have been focused on Jenkins ever since.

Reviewer 2: ```suggestion
I worked at Sun Microsystems for quite a long time on the NetBeans IDE.
We had some CI processes, such as Cruise Control, as part of that.
While there, I heard about Jenkins (then Hudson) in its infancy.
I knew of Kohsuke and his work on Hudson and thought it might be something we could start using.
At the time, we had been using CVS for version control, which Hudson had incorporated to some degree.

We migrated from CVS to Mercurial SCM, which I also wound up working on in Hudson.
It was just enough to get the features we needed for our usage patterns and we didn't pay more attention to it beyond that.
I had also done some work on the tool system where you can have tool installations and automatic downloads.
That was mostly driven by Andrew Bayer, another long-time Jenkins contributor.
After the Jenkins fork was created, Kohsuke contacted me and asked if I wanted to come join him at the company where he was working on Jenkins.
This was in 2012 when I joined CloudBees, and I have been focused on Jenkins ever since.
```

These requests also help drive architectural changes, despite the challenges they can present.
Such requests also help find and deal with technical debt.

== What contributions have felt the most successful or impactful?

I was eager to file one of the first substantive link:https://github.com/jenkinsci/jep/tree/master/jep[JEPs]: JEP-200.
The goal of this JEP was to greatly reduce the surface area of Jenkins threatened by deserialization “gadgets”.
The work planned was in the same spirit as the Jenkins CLI simplification, which predated the JEP system.
Another notable feature was JEP-222, allowing WebSocket to be used to connect agents to the controller.
This was inspired by the trickiness of setting up “level 4” network proxies, a common hurdle for CloudBees CI customers, especially in Kubernetes.
JEP-227 was a very tedious, yet long overdue, project to update to the (then) current Spring Security release across several major versions.
Preventing almost all user-visible regressions required a lot of refactoring in Jenkins core and a lot of plugin updates.
It is great to see that the Jenkins project has gotten much more proactive in updating dependencies since those days.

> At the infrastructure level, JEP-305 made it possible to prototype changes spanning multiple plugins, without sacrificing CI coverage, despite the lack of a “monorepo” for Jenkins development.
> More recently, JEP-229 built on that system to make it possible to publish plugin releases automatically, without requiring maintainers to keep Artifactory credentials on their laptop.
> This had been a longstanding source of frustrating mistakes and a security risk.
> Both changes were inspired especially by work on Pipeline-related plugins, which form a loose coalition without a fixed boundary: I wanted it to be as simple as possible to develop, test, and publish changes rapidly.

Reviewer 2: ```suggestion
At the infrastructure level, JEP-305 made it possible to prototype changes spanning multiple plugins, without sacrificing CI coverage, despite the lack of a “monorepo” for Jenkins development.
More recently, JEP-229 built on that system to make it possible to publish plugin releases automatically, without requiring maintainers to keep Artifactory credentials on their laptops.
This had been a longstanding source of frustrating mistakes and a security risk.
Both changes were inspired especially by work on Pipeline-related plugins, which form a loose coalition without a fixed boundary: I wanted it to be as simple as possible to develop, test, and publish changes rapidly.
```
